builder/common/ssm/session.go

```diff
@@ -84,21 +86,48 @@ func (s Session) getCommand(ctx context.Context) ([]string, string, error) {
 	return args, *session.SessionId, nil
 }
 
+// terminate an interactive Systems Manager session with a remote instance via the
+// AWS session-manager-plugin. Session cannot be resumed after termination.
+func (s Session) terminateSession(sessionID string, ui packersdk.Ui) {
+	log.Printf("ssm: Terminating PortForwarding session %q", sessionID)
+	_, err := s.SvcClient.TerminateSession(&ssm.TerminateSessionInput{SessionId: aws.String(sessionID)})
+	if err != nil {
+		ui.Error(fmt.Sprintf("Error terminating SSM Session %q, this does not affect the built AMI. Please terminate the session manually: %s", sessionID, err))
+	}
+}
+
 // Start an interactive Systems Manager session with a remote instance via the
 // AWS session-manager-plugin. To terminate the session you must cancell the
 // context. If you do not wish to terminate the session manually: calling
 // StopSession on a instance of this driver will terminate the active session
 // created from calling StartSession.
 // To stop the session you must cancel the context.
 func (s Session) Start(ctx context.Context, ui packersdk.Ui, sessionChan chan struct{}) error {
-	for ctx.Err() == nil {
+	exitSession := false
+	for ctx.Err() == nil && !exitSession {
 		log.Printf("ssm: Starting PortForwarding session to instance %s", s.InstanceID)
 		args, sessionID, err := s.getCommand(ctx)
 		if sessionID != "" {
 			defer func() {
-				_, err := s.SvcClient.TerminateSession(&ssm.TerminateSessionInput{SessionId: aws.String(sessionID)})
-				if err != nil {
-					ui.Error(fmt.Sprintf("Error terminating SSM Session %q. Please terminate the session manually: %s", sessionID, err))
+				s.terminateSession(sessionID, ui)
+			}()
+			// If the instance is terminated the session must be terminated as well
+			// Otherwise the start-session command will exit with status -1
+			go func() {
+				for {
+					instanceState, err := s.Ec2Conn.DescribeInstanceStatus(&ec2.DescribeInstanceStatusInput{
+						InstanceIds: []*string{aws.String(s.InstanceID)},
+					})
+					if err != nil {
+						log.Printf("ssm: Error describing instance status: %s", err)
+					}
+					// if no instance status is returned, the instance is terminated
+					if len(instanceState.InstanceStatuses) == 0 {
```

Comment:
In addition to this we could also probably terminate the session when the instance is being shut down, what do you think?

Reply:
Good suggestion

Reply:
I might be wrong, but if the `DescribeInstanceStatus` call fails, `instanceState` may be `nil`, right? If so, this will crash the plugin as deferencing a `nil` object causes a NRE, and since it's in a coroutine and it's never caught, the whole plugin will come crashing down if/when that happens.

We should check for that here.

Reply:
Also, in terms of permissions, I think it'll be rare (maybe impossible in our context), but if someone cannot perform the `DescribeInstanceStatus` call, this coroutine will never exit, will it?
This will also be true if for some reason the API call always ends with an error, as such, I would suggest we add a way to send a message from the outside, for example when `localexec.RunAndStream` exits naturally, this way we can stop checking periodically, and exit the coroutine when that happens.

Reply:
If we are waiting on outside events as well, I suggest we use a `time.Ticker` for the periodic checks. This way we can use a `select` statement, and can react immediately to whichever event triggers on a channel, and immediately quit if we do receive that message.

Reply:
Good point, thanks. I'm going to document the permission as well in the docs. In case someone gets that error again they can check if they have the permission.